internal/bigmap/bigmap_internal.go

```diff
@@ -0,0 +1,476 @@
+	defaultFileSegmentSize      = 1024 << 20 // 1 GiB
+	defaultInitialSizeLogarithm = 20
+	mmapFileMode                = 0o600
+
+	// grow hash table above this percentage utilization, higher values (close to 100) will be very slow,
+	// smaller values will waste memory.
+	defaultLoadFactorPercentage = 75
+
+	minKeyLength = 4 // first 4 bytes of each key are used as uint32 hash value
+	maxKeyLength = 255
+
+	minSizeLogarithm = 8
+)
+
+var log = logging.Module("bigmap")
+
+// Options provides options for the internalMap.
+type Options struct {
+	LoadFactorPercentage int   // grow the size of the hash table when this percentage full
+	NumMemorySegments    int   // number of segments to keep in RAM
+	MemorySegmentSize    int64 // size of a single memory segment
+	FileSegmentSize      int   // size of a single file segment, defaults to 1 GiB
+
+	InitialSizeLogarithm int // logarithm of the initial size of the hash table, default - 20
+}
+
+// internalMap is a custom hashtable implementation using https://en.wikipedia.org/wiki/Double_hashing
+// that stores all (key,value) pairs densely packed in segments of fixed size to minimize data
+// fragmentation.
+type internalMap struct {
+	hasValues bool    // +checklocksignore
+	opts      Options // +checklocksignore
+
+	mu sync.RWMutex
+	// +checklocks:mu
+	tableSizeIndex int // index into tableSizesPrimes == len(slots)
+	// +checklocks:mu
+	segments []mmap.MMap // ([keyLength][key][valueLength][value])*
+	// +checklocks:mu
+	slots []entry // current hash table slots
+
+	// +checklocks:mu
+	count int // number of elements in the hash tables
+	// +checklocks:mu
+	h2Prime uint64 // prime < len(slots)
+
+	// +checklocks:mu
+	cleanups []func()
+
+	// +checklocks:mu
+	tempDir string
+}
+
+// The list of prime numbers close to 2^N from https://primes.utm.edu/lists/2small/0bit.html
+
+//nolint:gochecknoglobals
+var tableSizesPrimes = []uint64{
+	1<<8 - 59, // minSizeLogarithm
+	1<<9 - 55,
+	1<<10 - 57,
+	1<<11 - 61,
+	1<<12 - 77,
+	1<<13 - 91,
+	1<<14 - 111,
+	1<<15 - 135,
+	1<<16 - 123,
+	1<<17 - 99,
+	1<<18 - 93,
+	1<<19 - 87,
+	1<<20 - 185, // initial H2 prime
+	1<<21 - 129, // initial number of hashtable slots
+	1<<22 - 123,
+	1<<23 - 159,
+	1<<24 - 167,
+	1<<25 - 183,
+	1<<26 - 135,
+	1<<27 - 235,
+	1<<28 - 273,
+	1<<29 - 133,
+	1<<30 - 173,
+	1<<31 - 171,
+	1<<32 - 267,
+	1<<33 - 355,
+	1<<34 - 281,
+	1<<35 - 325,
+	1<<36 - 233,
+	1<<37 - 375,
+	1<<38 - 257,
+	1<<39 - 301,
+	1<<40 - 437,
+	1<<41 - 139,
+	1<<42 - 227,
+	1<<43 - 369,
+	1<<44 - 377,
+	1<<45 - 229,
+	1<<46 - 311,
+	1<<47 - 649,
+	1<<48 - 257,
+	1<<49 - 339,
+	1<<50 - 233,
+	1<<51 - 439,
+	1<<52 - 395,
+	1<<53 - 421,
+	1<<54 - 327,
+	1<<55 - 267,
+	1<<56 - 195,
+	1<<57 - 423,
+	1<<58 - 251,
+	1<<59 - 871,
+	1<<60 - 453,
+	1<<61 - 465,
+	1<<62 - 273,
+	1<<63 - 471,
+}
+
+type entry struct {
+	segment uint32 // 0-empty, otherwise index into internalMap.segments+1
+	offset  uint32 // offset within a segment
```

Comment:
@jkowalski thinking about this...

Are the entries (key + value) expected to be of fixed size?

If so, given that this map is specialized for this use case, etc.,
Would it make sense to make the key and value sizes be parameters of the map, and replace `offset` with an "`index`" into the segment? The offset for the entry would then be computed using the entry size.

Thoughts?